src/v/storage/index_state.h

```diff
@@ -77,9 +116,37 @@ struct index_state
             non_data_timestamps = false;
         }
     }
-    std::tuple<uint32_t, uint32_t, uint64_t> get_entry(size_t i) {
+    std::tuple<uint32_t, offset_time_index, uint64_t> get_entry(size_t i) {
         return {
-          relative_offset_index[i], relative_time_index[i], position_index[i]};
+          relative_offset_index[i],
+          offset_time_index{relative_time_index[i]},
+          position_index[i]};
+    }
+
+    std::optional<std::tuple<uint32_t, offset_time_index, uint64_t>>
+    find_entry(model::timestamp ts) {
+        const auto idx = offset_time_index{ts};
+
+        auto it = std::lower_bound(
+          std::begin(relative_time_index),
+          std::end(relative_time_index),
+          idx.raw_value(),
+          std::less<uint32_t>{});
+        if (it == relative_offset_index.end()) {
+            return std::nullopt;
+        }
+
+        const auto dist = std::distance(relative_offset_index.begin(), it);
+
+        // lower_bound will place us on the first batch in the index that has
+        // 'max_timestamp' greater than 'ts'. Since not every batch is indexed,
+        // it's not guaranteed* that 'ts' will be present in the batch
+        // (i.e. 'ts > first_timestamp'). For this reason, we go back one batch.
+        //
+        // *In the case where lower_bound places on the first batch, we'll
+        // start the timequery from the beggining of the segment as the user
+        // data batch is always indexed.
+        return get_entry(std::max(dist - 1, decltype(dist){0}));
```

Comment:
nit:

instead of having decltype and depending on relative_offset_index::difference_type being signed and implicitly relying on underflow of `dist-1` for dist=0 case (i'm guess this last part is true), you could do: `get_entry(dist > 0 : dist - 1 : 0;)` ?